main.go

```diff
@@ -143,19 +165,73 @@ func getLocator(ctx context.Context) (*ipfsgeoip.IPLocator, error) {
 
 	go lite.Bootstrap(ipfslite.DefaultBootstrapPeers())
 
-	// An exchange session will speed things up
 	return ipfsgeoip.NewIPLocator(lite.Session(ctx)), nil
 }
 
-func getMinerInfos(db *pg.DB) ([]minerInfo, error) {
+func getRouter(ctx context.Context) (*FilecoinPeer, error) {
+	fp, err := SetupFilecoinPeer(
+		ctx,
+		"mainnet",
+	)
+
+	if err != nil {
+		return &FilecoinPeer{}, nil
+	}
+
+	fp.Crawl(crawlTime)
+
+	return fp, nil
+}
+
+func getMinerInfos(db *pg.DB, fp *FilecoinPeer) ([]minerInfo, error) {
 	// Tl;dr: get latest known miner_id, multi_addresses, height for each
 	// miner.
-	var infos []minerInfo
-	_, err := db.Query(&infos, query)
+	knownPeers := fp.host.Peerstore().Peers()
+
+	var infosFromDb []minerInfo
+	var infosFromFp []minerInfo = make([]minerInfo, len(knownPeers))
+
+	_, err := db.Query(&infosFromDb, query)
 	if err != nil {
 		return nil, err
 	}
-	return infos, nil
+	for _, info := range infosFromDb {
+		info.Source = "self-report"
+	}
+
+	var height int64
+	if _, err := db.Model((*minerInfo)(nil)).QueryOne(pg.Scan(&height), `
+SELECT height FROM ?TableName
+ORDER BY height DESC
+LIMIT 1
+`); err != nil {
+		log.Println("could not find most recent height")
+		log.Fatal(err)
+	}
+	for _, peer := range knownPeers {
+		peerAddrs := fp.host.Peerstore().PeerInfo(peer).Addrs
+		peerAddrsAsStr := make([]string, len(peerAddrs))
+		for _, p := range peerAddrs {
+			peerAddrsAsStr = append(peerAddrsAsStr, p.String())
+		}
+
+		info := minerInfo{
+			Height:         height,        // use most recent height from locations table for now
+			MinerID:        peer.String(), // cannot reliably get miner ID given a peer ID from miner_infos table
+			MultiAddresses: peerAddrsAsStr,
+		}
+
+		if len(info.MultiAddresses) == 0 { // sometimes multiaddr not found in peerstore
+			continue
+		}
+
+		info.Source = "routing"
+		infosFromFp = append(infosFromFp, info)
+	}
+
+	log.Println("COUNT INFOS FROM FP")
+	log.Println(len(infosFromFp))
+	return append(infosFromDb, infosFromFp...), nil
```

Comment:
I'm afraid this doesn't work. known_peers is going to be mostly the bootstrappers plus some nodes we connected to (thanks to the dht bootstrap that happens in the background).

Here I would expect a function that:

- Makes a list of "active" miners. I.e. miners making power claims  in the last two weeks (power_actor_claims)
- Joined with miner_infos to obtain their `peer_id`.
- Then, for each of those, a call to https://pkg.go.dev/github.com/libp2p/go-libp2p-kad-dht#IpfsDHT.FindPeer (with a few workers doing these calls in parallel and a relatively quick timeout).

I would separate these things on their own function so that it does not get mixed with current getMinerInfos() (can be renamed to getAddressesFromMinerInfos).

Reply:
Thanks! Addressed your feedback [here](https://github.com/filecoin-project/sentinel-locations/pull/2/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R200).